src/modules/Game.class.js

```diff
@@ -1,68 +1,263 @@
-   * @returns {number[][]}
-   */
-  getState() {}
-
-  /**
-   * Returns the current game status.
-   *
-   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
-   *
-   * `idle` - the game has not started yet (the initial state);
-   * `playing` - the game is in progress;
-   * `win` - the game is won;
-   * `lose` - the game is lost
-   */
-  getStatus() {}
-
-  /**
-   * Starts the game.
-   */
-  start() {}
-
-  /**
-   * Resets the game.
-   */
-  restart() {}
-
-  // Add your own methods here
+  static gameStatuses = {
+    idle: 'idle',
+    playing: 'playing',
+    win: 'win',
+    lose: 'lose',
+  };
+
+  constructor(
+    initialState = Array.from({ length: 4 }, () => Array(4).fill(0)),
+  ) {
+    this.initialState = initialState;
+    this.state = initialState.map((row) => [...row]);
+    this.status = Game.gameStatuses.idle;
+    this.score = 0;
+  }
+
+  moveLeft() {
+    if (!this.isStateValid(this.state)) {
+      return;
+    }
+
+    const updatedState = this.state.map((row) => this.applyMove(row));
+
+    this.updateGameState(updatedState);
+    this.completeMoveTasks();
+  }
+
+  moveRight() {
+    const reversedState = this.state.map((row) => [...row].reverse());
+
+    if (!this.isStateValid(reversedState)) {
+      return;
+    }
+
+    const updatedState = reversedState.map((row) =>
+      // eslint-disable-next-line prettier/prettier
+      this.applyMove(row).reverse());
+
+    this.updateGameState(updatedState);
+    this.completeMoveTasks();
+  }
+
+  moveUp() {
+    const rotatedRightState = this.rotateRight(this.state);
+
+    if (!this.isStateValid(rotatedRightState)) {
+      return;
+    }
+
+    const newState = [
+      ...rotatedRightState.map((row) => this.applyMove([...row])),
+    ];
+
+    const rotatedLeftState = this.rotateLeft(newState);
+
+    this.updateGameState(rotatedLeftState);
+    this.completeMoveTasks();
+  }
+
+  moveDown() {
+    const rotatedRightState = this.rotateRight(this.state);
+    const rotatedLocaleState = [
+      ...rotatedRightState.map((row) => [...row].reverse()),
+    ];
+
+    if (!this.isStateValid(rotatedLocaleState)) {
+      return;
+    }
+
+    const newState = rotatedLocaleState.map((row) =>
+      // eslint-disable-next-line prettier/prettier
+      this.applyMove([...row]).reverse());
+    const rotatedLeftState = this.rotateLeft(newState);
+
+    this.updateGameState(rotatedLeftState);
+    this.completeMoveTasks();
+  }
+
+  applyMove(vector) {
+    const newRow = [];
+
+    let i = 0;
+
+    while (i < vector.length) {
+      const current = vector[i];
+      const next = vector[i + 1];
+
+      if (current) {
+        if (current === next) {
+          newRow.push(current * 2);
+          this.score += current * 2;
+          i += 2;
+        } else {
+          newRow.push(current);
+          i++;
+        }
+      } else {
+        i++;
+      }
+    }
+
+    while (newRow.length < vector.length) {
+      newRow.push(0);
+    }
+
+    return newRow;
```

Comment:
The `applyMove` method does not handle the case where there are multiple pairs of mergeable tiles in a row. For example, if the input is `[2, 2, 2, 2]`, the expected output should be `[4, 4, 0, 0]`, but the current implementation will output `[4, 2, 0, 0]`. You need to adjust the loop to handle subsequent merges correctly.